File: sdk/swift/Sources/HiTechCloudSDK/BareMetal.swift
import Foundation

/// Bare Metal & Colocation: reinstall, diagnostics, rescue, server info, IPs, VLANs, power
public class BareMetalResource {
    private let httpClient: HTTPClient
    public init(httpClient: HTTPClient) { self.httpClient = httpClient }

    /// GET /api/service/{id}/reinstall/templates - List OS templates and recipes
    public func listReinstallTemplates(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/reinstall/templates", params: params)
    }

    /// POST /api/service/{id}/diag/cancel - Cancel diagnostics operation
    public func cancelDiagnostics(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/diag/cancel", params: params)
    }

    /// GET /api/service/{id}/diag/templates - List diagnostics templates
    public func listDiagnosticsTemplates(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/diag/templates", params: params)
    }

    /// GET /api/service/{id}/diag - Show diagnostic operation status
    public func getDiagnosticsStatus(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/diag", params: params)
    }

    /// POST /api/service/{id}/diag - Start new diagnostic operation
    public func runDiagnostics(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/diag", params: params)
    }

    /// GET /api/service/{id}/rescue/templates - List rescue templates
    public func listRescueTemplates(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/rescue/templates", params: params)
    }

    /// GET /api/service/{id}/rescue - Show rescue operation status
    public func getRescueStatus(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/rescue", params: params)
    }

    /// POST /api/service/{id}/rescue - Start rescue operation
    public func rescue(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/rescue", params: params)
    }

    /// POST /api/service/{id}/rescue/cancel - Cancel rescue operation
    public func cancelRescue(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/rescue/cancel", params: params)
    }

    /// GET /api/service/{id}/info - Get server hardware and OS details
    public func getServerInfo(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/info", params: params)
    }

    /// POST /api/service/{id}/hostname - Update server hostname
    public func updateHostname(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/hostname", params: params)
    }

    /// GET /api/service/{id}/ips - List server IPs
    public func listIps(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/ips", params: params)
    }

    /// POST /api/service/{id}/ips - Add new IP to server
    public func addIp(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/ips", params: params)
    }

    /// GET /api/service/{id}/vlans - List available VLANs
    public func listVlans(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/vlans", params: params)
    }

    /// GET /api/service/{id}/ips/{ip} - Get IP details
    public func getIpDetails(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/ips/{ip}", params: params)
    }

    /// PUT /api/service/{id}/ips/{ip} - Edit IP
    public func editIp(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.put("/api/service/{id}/ips/{ip}", params: params)
    }

    /// GET /api/service/{id}/reinstall - Return installation options
    public func getReinstallDetails(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/reinstall", params: params)
    }

    /// POST /api/service/{id}/reinstall - Reinstall server
    public func reinstall(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/reinstall", params: params)
    }

    /// GET /api/service/{id}/status - Return server power status
    public func getStatus(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/status", params: params)
    }

    /// POST /api/service/{id}/power/{action} - Power action (start/stop/reboot/reset/shutdown)
    public func powerAction(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/power/{action}", params: params)
    }

    /// GET /api/serverstock - List servers in stock
    public func listServersInStock(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/serverstock", params: params)
    }

}

File: sdk/kotlin/src/main/kotlin/vn/hitechcloud/sdk/BareMetal.kt
package vn.hitechcloud.sdk

/** Bare Metal & Colocation: reinstall, diagnostics, rescue, server info, IPs, VLANs, power */
class BareMetalResource(private val httpClient: HttpClient) {
    /** GET /api/service/{id}/reinstall/templates - List OS templates and recipes */
    fun listReinstallTemplates(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/reinstall/templates", params)
    }

    /** POST /api/service/{id}/diag/cancel - Cancel diagnostics operation */
    fun cancelDiagnostics(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/diag/cancel", params)
    }

    /** GET /api/service/{id}/diag/templates - List diagnostics templates */
    fun listDiagnosticsTemplates(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/diag/templates", params)
    }

    /** GET /api/service/{id}/diag - Show diagnostic operation status */
    fun getDiagnosticsStatus(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/diag", params)
    }

    /** POST /api/service/{id}/diag - Start new diagnostic operation */
    fun runDiagnostics(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/diag", params)
    }

    /** GET /api/service/{id}/rescue/templates - List rescue templates */
    fun listRescueTemplates(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/rescue/templates", params)
    }

    /** GET /api/service/{id}/rescue - Show rescue operation status */
    fun getRescueStatus(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/rescue", params)
    }

    /** POST /api/service/{id}/rescue - Start rescue operation */
    fun rescue(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/rescue", params)
    }

    /** POST /api/service/{id}/rescue/cancel - Cancel rescue operation */
    fun cancelRescue(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/rescue/cancel", params)
    }

    /** GET /api/service/{id}/info - Get server hardware and OS details */
    fun getServerInfo(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/info", params)
    }

    /** POST /api/service/{id}/hostname - Update server hostname */
    fun updateHostname(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/hostname", params)
    }

    /** GET /api/service/{id}/ips - List server IPs */
    fun listIps(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/ips", params)
    }

    /** POST /api/service/{id}/ips - Add new IP to server */
    fun addIp(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/ips", params)
    }

    /** GET /api/service/{id}/vlans - List available VLANs */
    fun listVlans(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/vlans", params)
    }

    /** GET /api/service/{id}/ips/{ip} - Get IP details */
    fun getIpDetails(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/ips/{ip}", params)
    }

    /** PUT /api/service/{id}/ips/{ip} - Edit IP */
    fun editIp(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.put("/api/service/{id}/ips/{ip}", params)
    }

    /** GET /api/service/{id}/reinstall - Return installation options */
    fun getReinstallDetails(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/reinstall", params)
    }

    /** POST /api/service/{id}/reinstall - Reinstall server */
    fun reinstall(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/reinstall", params)
    }

    /** GET /api/service/{id}/status - Return server power status */
    fun getStatus(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/status", params)
    }

    /** POST /api/service/{id}/power/{action} - Power action (start/stop/reboot/reset/shutdown) */
    fun powerAction(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/power/{action}", params)
    }

    /** GET /api/serverstock - List servers in stock */
    fun listServersInStock(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/serverstock", params)
    }

}
